core/src/main/java/org/elasticsearch/common/settings/Settings.java

```diff
@@ -1032,7 +1032,110 @@ public Builder normalizePrefix(String prefix) {
          * set on this builder.
          */
         public Settings build() {
-            return new Settings(Collections.unmodifiableMap(map));
+            return new Settings(map);
+        }
+    }
+
+    // TODO We could use an FST internally to make things even faster and more compact
+    private static final class FilteredMap extends AbstractMap<String, String> { // pkg private for testing
+        private final Map<String, String> delegate;
+        private final Predicate<String> filter;
+        private final String prefix;
+        private int size = -1; // we cache that size since we have to iterate the entire set
```

Comment:
maybe also add a comment that it is safe to do so since the underlying map is immutable? (I think it is based on other comments?)